scripts/sign_encrypt.py

```diff
@@ -488,10 +606,82 @@ def display(self):
                 if len(img) != img_size:
                     raise Exception("Unexpected img size: got {}, expected {}"
                                     .format(len(img), img_size))
-        else:
-            raise Exception("Unsupported image type: {}".format(img_type))
+            offs += img_size
 
 
+        offs = 0
+        while offs < len(self.inf):
+            if offs > 0:
+                # name_size is the previous subkey header
+                name_img = self.inf[offs:offs + name_size]
+                print('  next name:  "{}"'.format(name_img_to_str(name_img)))
+                offs += name_size
+                print('Next header at offset: {} (0x{:x})'
+                      .format(offs, offs))
+
+            shdr = self.inf[offs:offs + SHDR_SIZE]
+            [magic, img_type, img_size, algo_value, digest_len,
+             sig_len] = struct.unpack('<IIIIHH', shdr)
+            offs += SHDR_SIZE
+
+            if magic != SHDR_MAGIC:
+                Exception("Unexpected magic: 0x{:08x}".format(magic))
+
+            img_type_name ='Unknown'
+            if img_type == SHDR_BOOTSTRAP_TA:
+                print('Bootstrap TA')
+                img_type_name = 'SHDR_BOOTSTRAP_TA'
+            if img_type == SHDR_ENCRYPTED_TA:
+                print('Encrypted TA')
+                img_type_name = 'SHDR_ENCRYPTED_TA'
+            if img_type == SHDR_SUBKEY:
+                print('Subkey')
+                img_type_name = 'SHDR_SUBKEY'
+
+            algo_name ='Unknown'
+            if algo_value in sig_tee_alg.values():
+                algo_name = value_to_key(sig_tee_alg, algo_value)
+
+            print('  magic:      0x{:08x}'.format(magic))
+            print('  img_type:   {} ({})'.format(img_type, img_type_name))
+            print('  img_size:   {} bytes'.format(img_size))
+            print('  algo:       0x{:08x} ({})'.format(algo_value, algo_name))
+            print('  hash_size:  {} bytes'.format(digest_len))
+            print('  sig_size:   {} bytes'.format(sig_len))
+
+            if algo_value not in sig_tee_alg.values():
+                raise Exception('Unrecognized algorithm: 0x{:08x}'
+                                .format(algo_value))
+
+            if digest_len != self.digest_len:
+                    raise Exception("Unexpected digest len: {}"
+                                    .format(digest_len))
+
+            img_digest = self.inf[offs:offs + digest_len]
+            print('  hash:       {}...'
+                  .format(binascii.hexlify(img_digest[:12])))
```

Comment:
Why truncate?

Reply:
SUre but honestly having a decoding tool print slightly long data isn't unexpected, and 80 character is not that much. It might come in handy to troubleshoot corruption some day.
Then what about showing the start of `struct shdr` and`struct shdr_subkey`? It would help understand the fact that `algo` is present twice in the same block of data. BTW the algo is not decoded to a name the second time.
Also, drop the Python binary string prefix (b'xxx').
```diff
diff --git a/scripts/sign_encrypt.py b/scripts/sign_encrypt.py
index 1141920af..924c2f664 100755
--- a/scripts/sign_encrypt.py
+++ b/scripts/sign_encrypt.py
@@ -556,6 +556,7 @@ class Binary_image:
 
         def display_ta():
             nonlocal offs
+            print(' struct shdr_bootstrap_ta')
             ta_uuid = self.inf[offs:offs + UUID_SIZE]
             print('  uuid:       {}'.format(uuid.UUID(bytes=ta_uuid)))
             offs += UUID_SIZE
@@ -569,6 +570,7 @@ class Binary_image:
                 [enc_algo, flags, nonce_len,
                  tag_len] = struct.unpack('<IIHH', ehdr)
 
+                print(' struct shdr_encrypted_ta')
                 enc_algo_name = 'Unkown'
                 if enc_algo in enc_tee_alg.values():
                     enc_algo_name = value_to_key(enc_tee_algo, enc_algo)
@@ -600,8 +602,8 @@ class Binary_image:
                                     .format(len(ciphertext), img_size))
             else:
                 img = self.inf[offs:]
-                print('  TA offset:  {} (0x{:x}) bytes'.format(offs, offs))
-                print('  TA size:    {} (0x{:x}) bytes'
+                print(' TA offset:  {} (0x{:x}) bytes'.format(offs, offs))
+                print(' TA size:    {} (0x{:x}) bytes'
                       .format(len(img), len(img)))
                 if len(img) != img_size:
                     raise Exception("Unexpected img size: got {}, expected {}"
@@ -642,6 +644,7 @@ class Binary_image:
             if algo_value in sig_tee_alg.values():
                 algo_name = value_to_key(sig_tee_alg, algo_value)
 
+            print(' struct shdr')
             print('  magic:      0x{:08x}'.format(magic))
             print('  img_type:   {} ({})'.format(img_type, img_type_name))
             print('  img_size:   {} bytes'.format(img_size))
@@ -658,8 +661,8 @@ class Binary_image:
                                     .format(digest_len))
 
             img_digest = self.inf[offs:offs + digest_len]
-            print('  hash:       {}...'
-                  .format(binascii.hexlify(img_digest[:12])))
+            print('  hash:       {}'
+                  .format(binascii.hexlify(img_digest).decode('ascii')))
             offs += digest_len
             sig = self.inf[offs:offs + sig_len]
             offs += sig_len
@@ -672,6 +675,7 @@ class Binary_image:
                                       offs + UUID_SIZE + SK_HDR_SIZE]
                 [name_size, subkey_version, max_depth, algo,
                  attr_count] = struct.unpack('<IIIII', img_subkey)
+                print(' struct shdr_subkey')
                 print('  uuid:       {}'.format(uuid.UUID(bytes=img_uuid)))
                 print('  name_size:  {}'.format(name_size))
                 print('  subkey_version: {}'.format(subkey_version))
```
```
Subkey
 struct shdr
  magic:      0x4f545348
  img_type:   3 (SHDR_SUBKEY)
  img_size:   320 bytes
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  hash_size:  32 bytes
  sig_size:   256 bytes
  hash:       0386b87b5832e571752bd96c9db8cfade9807db2b926f8529fcd1953362cc5d2
 struct shdr_subkey
  uuid:       fd02c9da-306c-48c7-a49c-bbd827ae86ee
  name_size:  0
  subkey_version: 0
  max_depth:  2
  algo:       0x70414930
  attr_count: 2
  next name:  ""
Next header at offset: 628 (0x274)
Subkey
 struct shdr
  magic:      0x4f545348
  img_type:   3 (SHDR_SUBKEY)
  img_size:   320 bytes
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  hash_size:  32 bytes
  sig_size:   256 bytes
  hash:       c774cc6883367487e5a94c2263fab556984c09d30ee1a7cb95f1ac8f4940eed9
 struct shdr_subkey
  uuid:       fd02c9da-306c-48c7-a49c-bbd827ae86ee
  name_size:  0
  subkey_version: 0
  max_depth:  1
  algo:       0x70414930
  attr_count: 2
  next name:  ""
Next header at offset: 1256 (0x4e8)
Bootstrap TA
 struct shdr
  magic:      0x4f545348
  img_type:   1 (SHDR_BOOTSTRAP_TA)
  img_size:   222064 bytes
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  hash_size:  32 bytes
  sig_size:   256 bytes
  hash:       409905657c6168257ba85d98f0e2529e1672c34954345416cf43a9a0d04051ca
 struct shdr_bootstrap_ta
  uuid:       fd02c9da-306c-48c7-a49c-bbd827ae86ee
  ta_version: 0
 TA offset:  1584 (0x630) bytes
 TA size:    222064 (0x36370) bytes
```